File: typescript/src/module7/godclass/after/ReleaseNotifier.ts
import { requireNonNull } from '../../../shared/requireNonNull.js';
import type { PublishedRelease } from '../PublishedRelease.js';

export class ReleaseNotifier {
  readonly #notifications: string[] = [];
  private readonly eventSink: (event: string) => void;

  constructor(eventSink: (event: string) => void = () => { }) {
    this.eventSink = requireNonNull(eventSink, 'eventSink');
  }

  notifyPublished(release: PublishedRelease): void {
    requireNonNull(release, 'release');
    this.#notifications.push('release-published:' + release.releaseId);
    this.eventSink('notify:' + release.releaseId);
  }

  notifications(): readonly string[] {
    return Object.freeze([...this.#notifications]);
  }
}

File: csharp/src/Training.Module7/GodClass/After/ReleaseNotifier.cs
namespace Training.Module7.GodClass.After;

public sealed class ReleaseNotifier
{
    private readonly List<string> _notifications = [];
    private readonly Action<string> _eventSink;

    public ReleaseNotifier()
        : this(_ => { })
    {
    }

    public ReleaseNotifier(Action<string> eventSink)
    {
        ArgumentNullException.ThrowIfNull(eventSink);
        _eventSink = eventSink;
    }

    public void NotifyPublished(PublishedRelease release)
    {
        ArgumentNullException.ThrowIfNull(release);
        _notifications.Add("release-published:" + release.ReleaseId);
        _eventSink("notify:" + release.ReleaseId);
    }

    public IReadOnlyList<string> Notifications() => _notifications.ToList().AsReadOnly();
}
